test/grafter/rdf/sparql_test.clj

```diff
@@ -69,7 +82,26 @@
               "?s ?p ?o ."
               "}")
          (#'sparql/rewrite-values-clauses q4 { [:s :p] [[(URI. "http://s1") (URI. "http://p1")]
-                                                      [(URI. "http://s2") (URI. "http://p2")]]}))))
+                                                      [(URI. "http://s2") (URI. "http://p2")]]})))
+    (is (same-query?
+          (str "SELECT * WHERE {"
+               "VALUES (?s ?p) { (UNDEF UNDEF) }"
+               "?s ?p ?o ."
+               "}")
+          (#'sparql/rewrite-values-clauses q4 { [:s :p] ::sparql/undef})))
```

Comment:
I've had to think about this one a bit, and @RicSwirrl helped me clarify the thinking on it.  But I think we should not allow use of `::sparql/undef` like this, as it's semantically at the wrong level of nesting.  

I think the intention here though is good, which is to allow the setting of a default behaviour for when you want to override an existing default binding.  

Better behaviours would be:

If you have a query of:

`SELECT * WHERE { VALUES ?s { UNDEF } ?s ?p ?o }`

Which is semantically equivalent to `select ?s ?p ?o ...`

Then providing a binding of `{:s []}` would effectively rewrite to `SELECT * WHERE { VALUES ?s {} ?s ?p ?o }`.  Likewise a `nil` `:s` should do the same (though the first expresses the intent more clearly).

If you want the default behaviour (specified in the query) then you simply don't provide an `:s` in your bindings.

Reply:
I'm confused. In your example above you seem to suggest that we would like to provide a binding of `{:s []}`

Should result in:

```sparql
SELECT * WHERE { 
  VALUES ?s {}
  ?s ?p ?o 
}
```
But the above is not allowed, it must be 

```sparql
SELECT * WHERE { 
  VALUES ?s { UNDEF }
  ?s ?p ?o 
}
```
... which is what we currently have.

Reply:
The syntax should be supported by stardog (at least 5.x) and definitely the SPARQL standard.  It's the equivalent of a sub select returning no results.